src/sync/app.hpp

```diff
@@ -58,6 +57,181 @@ class App {
         realm::util::Optional<uint64_t> default_request_timeout_ms;
     };
 
+    /**
+     * A struct representing a user API key as returned by the App server.
+    */
+    struct UserAPIKey {
+        // The ID of the key.
+        realm::ObjectId id;
+
+        // The actual key. Will only be included in
+        // the response when an API key is first created.
+        std::string key;
+
+        // The name of the key.
+        std::string name;
+
+        // Whether or not the key is disabled.
+        bool disabled;
+    };
+
+    /**
+     * A client for the user API key authentication provider which
+     * can be used to create and modify user API keys. This
+     * client should only be used by an authenticated user.
+    */
+    class UserAPIKeyProviderClient {
```

Comment:
How are these methods supposed to be exposed in the SDK's? I haven't seen any mention of them before I think?

Reply:
A lot of these methods also seem to only work if you are an admin user. Is that correct? Perhaps clarify for each method?

Reply:
Where are you seeing that they only work as an admin user? I do not believe that is the case.

> How are these methods supposed to be exposed in the SDK's? I haven't seen any mention of them before I think?

I left these out mistakenly– to some extent, I was originally thinking a lot of these methods would be on the `User` class (similar to how Realm initially had it), but that created too much bloat.

Ideally, these methods would be exposed the same way in the SDKs as they are here. A generic function or field on the `RealmApp` class that exposes the various provider clients.

Reply:
Maybe I just misunderstood exactly what this is. This is extra login options a user already logged in can configure then?

Reply:
https://docs.mongodb.com/stitch/authentication/api-key/

A Stitch Admin can configure this provider from the backend.

If enabled, a non-anonymous user can effectively generate "login" keys and then login "elsewhere" with them. 

```cpp
/// from main relay
app.provider_client<UseAPIKeyProviderClient>().create_api_key(
    "my_api_key",
    [=](std::shared_ptr<UserAPIKey> api_key, Optional<AppError> error) {
        if (error) {
            ...
        }

        iot_device.send_api_key(api_key.key);
    }
);

// from iot_device
void send_api_key(const std::string& key) {
    app.login_with_credential(AppCredentials::UserAPIKey(key), ...); 
}
```
This would be common amongst IoT users, or anyone doing emulating a primary/replica pattern.